lib/config.js

```diff
@@ -352,11 +352,17 @@ const CONFIG_SYNTAX_HELP = '  module.exports = function(config) {\n' +
   '  };\n'
 
 function parseConfig (configFilePath, cliOptions, parseOptions) {
+  const promiseConfig = parseOptions && parseOptions.promiseConfig === true
```

Comment:
Should this change to:

```js
const throwErrors = parseOptions && parseOptions.throwErrors === true
const promiseConfig = throwErrors && parseOptions && parseOptions.promiseConfig === true
```
to have the rest of the function behave the same way as `fail()`?

Reply:
No, why?

`throwErrors` chooses between `process.exit()` and throw/reject.  
`promiseConfig` decides whether the function returns synchronously or returns a Promise.

Reply:
I understand (and mostly support) the logic of option A (`throwErrors`) controls process exiting and option B (`promiseConfig`) controls sync vs async.

My concern is that behavior would be surprising in the following scenario:
`{throwErrors: false, promiseConfig: true}`

-   (expected) critical/internal failures exit the process
-   (expected) non-critical failures return rejected promises
-   (unexpected, multiple behaviors) function exported by the config file:
    -   exceptions thrown are treated as critical failures and exit the process
    -   if the promise is rejected, we return that rejected promise instead of exiting the process

<details>

Other branches for reference:

-   `if (promiseConfig === false && configFileReturnsPromise === true)`
    -   critical failure (side thought: **should returned promises be silently ignored?** The existing sync behavior ignores it anyway. My current code tries to detect intent, which may be feature creep.)
-   `if (promiseConfig === true && configFileReturnsPromise === true)`
    -   success results in a promise that resolves to a `Config` instance
    -   if the promise returned by the config file is rejected, that rejection is carried through to the promise that `parseConfig` returns
-   `else if(promiseConfig === true && configFileReturnsPromise === false)`
    -   config is finalized and a resolved promise is returned with the `Config` instance as the fulfilled value.
    -   any errors thrown in `finalizeConfig` (and by extension `normalizeConfig`) are returned as a promise rejected with the exception as the reason
-   `if (promiseConfig === false && configFileReturnsPromise === false)`
    -   normal/existing sync behavior

</details>

I feel like the function exported by the config file should be treated one way or the other, but not both.

---

**Option 1**: My first approach to this was to have critical failures return a rejected promise with `promiseConfig: true` regardless of the value of `throwErrors`. This way thrown exceptions and rejected promises both return a rejected promise. It is up to the consumer of the public API to catch that rejection. Additionally, it made sense to me (personally) to not exit the process when using a method that you expect to return a promise.

This was rejected in a previous code review because one option (`promiseConfig`) implied the other (`throwErrors`) and it didn't seem like a clean solution.

---

**Option 2**: One way of approaching this is detecting `{throwErrors: false, promiseConfig: true}` and critically failing with a message stating the `throwsErrors` option must be `true` in order to use `promiseConfig`.

---

**Option 3**: Another approach could be to catch the promise rejection internally and exit the process, using the rejection reason to build the failure message.

---

**Option 4**: Leave the code as is

---

I'd personally like to avoid option 4, but I also don't want it to jeopardize the rest of the PR as it otherwise works well and the implementation is expected to change in the next major release anyway.

Reply:
I think we should go with option 3. The only change we need to do is below and that everything is consistent, right?

```diff
    return configModuleReturn.then(
      () => finalizeConfig(config),
      (e) => fail('Error in config file!\n', e)
    )
```

Reply:
Oh, and the below as well, to capture the `promiseConfig == true` with the sync config function.

```
try {
        return Promise.resolve(finalizeConfig(config))
      } catch (e) {
        return fail('Error in config file!\n', e)
      }
```